compiler/src/org.graalvm.compiler.lir.aarch64/src/org/graalvm/compiler/lir/aarch64/AArch64ArrayIndexOfOp.java

```diff
@@ -77,190 +90,301 @@ public AArch64ArrayIndexOfOp(int arrayBaseOffset, JavaKind valueKind, LIRGenerat
+         * Searching the array sequentially for the target value. Note that all indexing is done via
+         * a negative offset from the first position beyond the end of search region.
          */
-        masm.sub(64, curIndex, arrayLength, fromIndex);
-        masm.compare(64, curIndex, chunkSize / charSize);
-        masm.branchConditionally(ConditionFlag.GE, searchByChunk);
-
+        Register endIndex;
+        if (findTwoConsecutive) {
+            /* The end of the region is the second to last element in the array */
+            endIndex = asRegister(temp5);
+            masm.sub(64, endIndex, arrayLength, 1);
+            /* adjust search length accordingly */
+            masm.sub(64, searchLength, searchLength, 1);
+        } else {
+            /* the end of the region is the last element in the array */
+            endIndex = arrayLength;
+        }
         /*
-         * Search sequentially for small strings. Note that all indexing is done via a negative
-         * offset from the first position beyond the end of the array.
+         * Set the base address to be at the end of the region baseAddress + (endIndex << shift
+         * size).
          */
-        masm.mov(64, endIndex, arrayLength);
-        // Set the base address to be at the end of the string, after the last char
-        masm.add(64, baseAddress, baseAddress, arrayLength, ShiftType.LSL, shiftSize);
-        // Initial index is (fromIndex - arrayLength) << shift size
+        masm.add(64, baseAddress, baseAddress, endIndex, ShiftType.LSL, shiftSize);
+        /* Initial index is -searchLength << shift size */
+        Register curIndex = searchLength;
         masm.sub(64, curIndex, zr, curIndex, ShiftType.LSL, shiftSize);
-        // Loop doing char-by-char search
-        masm.bind(searchByCharLoop);
-        int charBitSize = charSize * Byte.SIZE;
-        masm.ldr(charBitSize, curChar, AArch64Address.createRegisterOffsetAddress(charBitSize, baseAddress, curIndex, false));
-        masm.cmp(32, curChar, searchChar);
+        /* Loop doing element-by-element search */
+        masm.bind(searchByElementLoop);
+        masm.ldr(memAccessSize, curValue, AArch64Address.createRegisterOffsetAddress(memAccessSize, baseAddress, curIndex, false));
+        masm.cmp(compareSize, searchValue, curValue);
         masm.branchConditionally(ConditionFlag.EQ, match);
 
         /*
-         * Add charSize to the curIndex, and retry if the end of the array has not yet been reached
-         * (i.e. the curIndex is still < 0).
+         * Add elementSize to the curIndex and retry if the end of the search region has not yet
+         * been reached, i.e., the curIndex is still < 0.
          */
-        masm.adds(64, curIndex, curIndex, charSize);
-        masm.branchConditionally(ConditionFlag.MI, searchByCharLoop);
+        masm.adds(64, curIndex, curIndex, elementByteSize);
+        masm.branchConditionally(ConditionFlag.MI, searchByElementLoop);
```

Comment:
I'm now playing with the end index so that this check can (I think) be used in both the 1 and 2 element cases